### ISSUE TITLE
Crashes if tag only exists in remote repository (and not in local repository)

### ISSUE THREAD

**Issue by george-miller** [kubernetes/git-sync#172]
## Context
I want to use git-sync to do tag based deploys.  My desired behavior is whenever a tagged commit is pushed to the master branch, git-sync be up to date with that commit.  

## Current Behavior
I currently have it working such that the `GIT_SYNC_REV` environment variable gets updated through Jenkins whenever a tagged commit is pushed.  The issue comes up when the container restarts and crashes with (I redacted info and placed it in `<>`)
```
I0617 19:07:40.024518       1 main.go:487] cloned <repo_url>
I0617 19:07:40.024548       1 main.go:597] run("/tmp/git"): git rev-parse <new_tag>
E0617 19:07:40.056606       1 main.go:294] error syncing repo: error running command: exit status 128: "<new_tag>\nfatal: ambiguous argument '<new_tag>': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\n"
```

The reason this happens is that since the repository is already cloned, git-sync does not reclone it or even pull it before trying to do a `rev-parse` on the local repository.  It wants to do a diff between the local hash and the remote hash but it can't because the hash doesn't exist locally.

git-sync needs be able to *allow revs to be nonexistant locally*.  As long as they exist on the remote, we can pull them and update.

I can make a PR for these changes, it should be quite simple and go in the `hashForRev` and `syncRepo` functions.  Seem ok?

**Comment by thockin** [comment]
In theory I am OK with it.  Is the idea that you want to use "whatever tag is latest"? Or latest on a give branch?  Or something else?  Do you intend to restart git-sync every time that changes?

**Comment by Typositoire** [comment]
This is currently kicking us in the bum. Any idea when this will be fixed? I do have a hackish fix but I'd prefer a solid one :p

**Comment by thockin** [comment]
I am confused about the use case, and did not get answers.  Maybe you can answer @Typositoire ?

Env variables and args can not be edited on a pod, so to change to a new tag you must be creating a new pod, right?

Creating a new pod should always clone with the right ref, right?

How do you get into a situation where you have a repo already cloned and it doesn't have the ref you wanted?

**Comment by Typositoire** [comment]
We're using EFS as the volume which means data is not lost on re-create. We
need that repo to be shared across multiple pods and always the same thus
only one gitsync -> efs <- PODS

**Comment by thockin** [comment]
If you have multiple pods sharing one EFS-backed volume, how many git-sync
replicas do you have?  More than 1 and they will potentially fight with
each other,  Exactly 1 and you should not have this problem.

**Comment by Typositoire** [comment]
I only have one. The problem is not that. You can't rev parse a TAG before actually fetching it. Once fetched (i tried hotfixing it :P) local == remote so no update is done.

If you clone a repo before a tag is created and you keep the volume when you change the rev to fetch from let's say `v1` to `v2` the repo exist, so it doesn't clone and try to do `rev parse` on `v1` but it hard failed with something like `ambiguous "v1" blablabla`

**Comment by Typositoire** [comment]
Hope the issue is more clear. Depending on your timezone we can have a quick voice chat and I'll explain/show you in details I'm setup to replicate it at will.

**Comment by thockin** [comment]
Sorry, I am still lost.

git-sync requires the clone and first sync to work.  To do that, we run:

```
git clone --no-checkout 
git rev-parse
git worktree add
git git reset --hard
```

We are now certain that we have the tag locally.  Subsequently, we call

```
git ls-remote
```

...to check if it has changed.  (BTW, I am working on some changes to how this works, so I really need to understand your use case.

> If you clone a repo before a tag is created and you keep the volume when you change the rev to 
> fetch from let's say v1 to v2 the repo exist, so it doesn't clone and try to do rev parse on v1 but it 
> hard failed with something like ambiguous "v1" blablabla

I see - you are restarting the pod with new flags (new tag name) but keeping the repo volume. Hence you are actually starting up without a fresh clone (as if it crashed and recovered) and the tag really doesn't exist locally.  That is not something I designed for.  I will have to come up with an e2e case that covers it.  If you have time to look at that (test_e2e.sh is not complicated), it would help ensure I get the use-case modeling your own reality.

**Comment by Typositoire** [comment]
Yes, I kind of get that this is not something you accounted for :p If it's out of scope of git-sync I can totally understand, otherwise you can check here this is what I've done so far to "make it work" this is far from being a good solution but gives you pointers.

https://github.com/Typositoire/git-sync/tree/tag_handling

I'll take a look at test_e2e.sh

**Comment by george-miller** [comment]
`you are restarting the pod with new flags (new tag name) but keeping the repo volume` this is exactly correct and was my problem.  Thanks for hashing it out, sorry I was mia for a while there.

We're not actually doing this anymore but I still think it is a nice to have.  btw @thockin I see you all over kube repos doing so much for the community, you're amazing!  Just wanted to say thank you so much for your awesome work, it's like meeting a celebrity :D

**Comment by p0tr3c** [comment]
We have run into this when trying to clone repo with `depth` 1 on specific revision. This behaviour is possible with `git` tool so this would be very useful. Did anyone get to a PR state?

**Comment by thockin** [comment]
I think this shakes out of a major rework I have in flight.  The
problem is that I don't know when that will land (squeezed for time to
work on it).  If someone figures out sa more surgical fix, I am open
to it.

**Comment by esakarya** [comment]
anyone can share jenkins sample with kubernetes ?

**Comment by thockin** [comment]
Will be fixed in v4 (after #691 and followups)
